billing: Improve invoice plans as needed error handling.

laurynmm opened the pull request:
Updates `invoice_plans_as_needed` so that we track customer IDs where we failed to process the invoice due to an error. This allows us to continue invoicing other customers, instead of stopping the process at the first  failure. And it allows us to not invoice any other plans that may be connected to a customer that already had a failure.

**Notes**:
- First two prep commits, pull out the logic for sending internal billing email notices for fixed-price plan renewal reminders and for stale audit log data causing overdue invoices.
- The third prep commit, pulls out the logic for skipping `invoice_plan` for self-hosted customers when the remote server audit log data is stale.
- The final commit adds the error logging to `invoice_plans_as_needed`.

**Tests**:
- There are a few more `# nocoverage` lines added with these changes, which we might want to add coverage for before merging these changes.

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

laurynmm commented on corporate/lib/stripe.py:
```diff
@@ -5413,6 +5413,31 @@ def get_plan_renewal_or_end_date(plan: CustomerPlan, event_time: datetime) -> da
+) -> None:
+    # We expect to have both an end date and next invoice date
+    # for this CustomerPlan.
+    assert plan.end_date is not None
```
This might be better handled as raising a `BillingError`. Previously, a fixed-price plan without an end date wouldn't cause any issues here.

timabbott replied:
Yeah, that seems reasonable.

laurynmm commented on corporate/lib/stripe.py:
```diff
@@ -5504,37 +5533,20 @@ def invoice_plans_as_needed(event_time: datetime | None = None) -> None:
+        try_to_invoice_plan = True
         if remote_server:
-            free_plan_with_no_next_plan = (
-                not plan.is_a_paid_plan() and plan.status == CustomerPlan.ACTIVE
```
Switched this in the new helper function to use `is_complimentary_access_plan`, which didn't exist until a few months ago. But using the new function, it makes it clearer why we still process an invoice for this case.

laurynmm commented on corporate/lib/stripe.py:
```diff
@@ -5515,38 +5515,71 @@ def check_remote_server_audit_log_data(
+                "Invoicing failed due to AssertionError: Customer.id: %s, CustomerPlan.id: %s",
+                plan.customer.id,
+                plan.id,
+            )
```
I wasn't sure if we wanted a `except Exception as e:` here as well to catch any other unexpected and otherwise unhandled errors.

timabbott replied:
Probably not essential that we do so, but it would be reasonable.

timabbott replied:
A plausible good pattern is to just catch all exceptions and use `isinstance` checks for logic that wants to vary with the exception class name.

laurynmm replied:
That makes sense. Revised for that pattern.

laurynmm commented:
**Updates**:
- Updated to log all `Exceptions` and do `isinstance` checks for the actual logging content.
- Added a check for the plan end date before calling the helper function for the fixed-price plan renewal email. That's basically aligned with what we do now, so there's no change in this potentially causing errors.

timabbott commented:
Looks great, merged, thanks @laurynmm!
